lib/experimental/auto-inserting-blocks.php

```diff
@@ -0,0 +1,183 @@
+				// location) to that array.
+				array_push( $block['innerContent'], null );
+			}
+			return $block;
+		}
+
+		$anchor_block_index = array_search( $anchor_block, array_column( $block['innerBlocks'], 'blockName' ), true );
+		if ( false !== $anchor_block_index && ( 'after' === $relative_position || 'before' === $relative_position ) ) {
+			if ( 'after' === $relative_position ) {
+				$anchor_block_index++;
+			}
+			array_splice( $block['innerBlocks'], $anchor_block_index, 0, array( $inserted_block ) );
+
+			// Find matching `innerContent` chunk index.
+			$chunk_index = 0;
+			while ( $anchor_block_index > 0 ) {
+				if ( ! is_string( $block['innerContent'][ $chunk_index ] ) ) {
+					$anchor_block_index--;
+				}
+				$chunk_index++;
+			}
+			// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
+			// when rendering blocks, we also need to insert a value (`null`, to mark a block
+			// location) into that array.
+			array_splice( $block['innerContent'], $chunk_index, 0, array( null ) );
+		}
+		return $block;
+	};
+}
+
+function gutenberg_register_auto_inserted_blocks( $settings, $metadata ) {
+	if ( ! isset( $metadata['__experimentalAutoInsert'] ) ) {
+		return $settings;
+	}
+	$auto_insert = $metadata['__experimentalAutoInsert'];
+
+	$property_mappings = array(
+		'before'     => 'before',
+		'after'      => 'after',
+		'firstChild' => 'first_child',
+		'lastChild'  => 'last_child',
+	);
+
+	$inserted_block_name = $metadata['name'];
+	foreach ( $auto_insert as $anchor_block_name => $position ) {
+		// Avoid infinite recursion (auto-inserting into self).
+		if ( $inserted_block_name === $anchor_block_name ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Cannot auto-insert block next to itself.', 'gutenberg' ),
+				'6.4.0'
+			);
+			continue;
+		}
+
+		if ( ! isset( $property_mappings[ $position ] ) ) {
+			continue;
+		}
+
+		$mapped_position = $property_mappings[ $position ];
+
+		$inserted_block = array(
+			'blockName'    => $inserted_block_name,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+
+		// TODO: In the long run, we'd likely want some sort of registry for auto-inserted blocks.
+
+		// Auto-insert sibling and child blocks into the editor (via the templates and patterns
+		// REST API endpoints), and auto-insert sibling blocks on the frontend.
+		$inserter = gutenberg_auto_insert_block( $anchor_block_name, $mapped_position, $inserted_block );
+		add_filter( 'gutenberg_serialize_block', $inserter, 10, 1 );
+
+		$settings['auto_insert'][ $anchor_block_name ] = $mapped_position;
+	}
+
+	return $settings;
+}
+add_filter( 'block_type_metadata_settings', 'gutenberg_register_auto_inserted_blocks', 10, 2 );
+
+function gutenberg_parse_and_serialize_block_templates( $query_result ) {
+	foreach ( $query_result as $block_template ) {
+		if ( 'theme' !== $block_template->source ) {
```

Comment:
Any particular reason to limit sources to `theme`?  I had to comment this out for `plugin` sources to work.

Reply:
No particular reason -- the main thing here is that we only want to auto-insert if the template doesn't have any user modifications, and I simply didn't think of plugin-supplied file templates.

Changing this to 

```suggestion
		if ( 'custom' === $block_template->source ) {
```

should do the trick, I believe.